smart-hadoop-support/smart-hadoop/src/main/java/org/smartdata/hdfs/scheduler/SmallFileScheduler.java

```diff
@@ -0,0 +1,512 @@
+          actionInfo.setResult(
+              "Failed to get file states of small files. " + e.toString());
+          return ScheduleResult.FAIL;
+        }
+
+        // Check if the state of small file is NORMAL
+        if (fileStateMap.size() != 0) {
+          for (String smallFile : smallFileList) {
+            if (fileStateMap.containsKey(smallFile)) {
+              FileState.FileType smallFileType = fileStateMap.get(
+                  smallFile).getFileType();
+              if (smallFileType != FileState.FileType.NORMAL) {
+                String errMsg = String.format(
+                    "%s has invalid file state %s for small file compact.",
+                    smallFile, smallFileType.toString());
+                LOG.debug(errMsg);
+                actionInfo.setResult(errMsg);
+                return ScheduleResult.FAIL;
+              }
+            }
+          }
+        }
+
+        try {
+          // Lock container file and set whether exists already
+          // Lock small files
+          boolean exist = dfsClient.exists(containerFilePath);
+          containerFileLock.put(containerFilePath, exist);
+          compactSmallFileLock.addAll(smallFileList);
+        } catch (IOException e) {
+          String errMsg = String.format(
+              "Failed to check if the container file is exists: %s for %s.",
+              containerFilePath, e.toString());
+          LOG.error(errMsg);
+          actionInfo.setResult(errMsg);
+          return ScheduleResult.FAIL;
+        }
+      }
+
+      return ScheduleResult.SUCCESS;
+    } else if (UNCOMPACT_ACTION_NAME.equals(actionInfo.getActionName())) {
+      // Get container file and small file list of this action
+      String containerFilePath = actionInfo.getArgs().get(
+          SmallFileCompactAction.CONTAINER_FILE);
+
+      // Check if container file is locked
+      if (!containerFileLock.containsKey(containerFilePath)) {
+        // Get small files of the container file
+        List<String> smallFileList;
+        try {
+          smallFileList = metaStore.getSmallFilesByContainerFile(containerFilePath);
+        } catch (MetaStoreException e) {
+          String errMsg = String.format(
+              "Failed to get small files of the container file: %s for %s.",
+              containerFilePath, e.toString());
+          LOG.error(errMsg);
+          actionInfo.setResult(errMsg);
+          return ScheduleResult.FAIL;
+        }
+
+        if (smallFileList != null && !smallFileList.isEmpty()) {
+          // Check if small file path is valid and unlocked
+          for (String smallFile : smallFileList) {
+            if (smallFile == null || smallFile.isEmpty()) {
+              LOG.debug("Illegal small file path: " + smallFile);
+              actionInfo.setResult("Illegal small file path: " + smallFile);
+              return ScheduleResult.FAIL;
+            } else if (uncompactSmallFileLock.contains(smallFile)) {
```

Comment:
Any consideration to have `uncompactSmallFileLock`? As `smallFile` was got from container file, is it possible to get a same small file from different container files?